### PR TITLE
Add support for STM32F407x MCUs.

### PR DESCRIPTION
**Pull request by thpoll83**
Add support for STM32F407xE (512KB) and STM32F407xG (1MB) MCUs.

## Description

Extended the build and python files to support these two STM32F407 MCU variations using the ST_STM32F4_DISCOVERY (STM32F407xG) board files. The mcuconf.h already existed in ChibiOS.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ x] My change requires a change to the documentation.
- [ x] I have updated the documentation accordingly.
- [ x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/mcu_selection.mk [hunk 1]
```diff
@@ -288,6 +288,78 @@ ifneq ($(findstring STM32F401, $(MCU)),)
...
+
+  # Board: it should exist either in <chibios>/os/hal/boards/,
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= GENERIC_STM32_F407VG
+
+  USE_FPU ?= yes
+
+  # Options to pass to dfu-util when flashing
+  DFU_ARGS ?= -d 0483:DF11 -a 0 -s 0x08000000:leave
+  DFU_SUFFIX_ARGS ?= -v 0483 -p DF11
```
Comment (fauxpark) [inline quantum/mcu_selection.mk:321]:
```suggestion
```

File: quantum/mcu_selection.mk [hunk 2]
```diff
@@ -288,6 +288,78 @@ ifneq ($(findstring STM32F401, $(MCU)),)
...
```
Comment (fauxpark) [inline quantum/mcu_selection.mk:357]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: data/schemas/keyboard.jsonschema [hunk 3]
```diff
@@ -25,7 +25,7 @@
         },
         "processor": {
             "type": "string",
-            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "MK66F18", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F411", "STM32F446", "STM32G431", "STM32G474", "STM32L433", "STM32L443", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", "attiny85", "unknown"]
+            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "MK66F18", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F407XE", "STM32F407XG", "STM32F411", "STM32F446", "STM32G431", "STM32G474", "STM32L433", "STM32L443", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", "attiny85", "unknown"]
```
Comment (tzarc) [inline data/schemas/keyboard.jsonschema:28]:
```suggestion
            "enum": ["cortex-m0", "cortex-m0plus", "cortex-m3", "cortex-m4", "MKL26Z64", "MK20DX128", "MK20DX256", "MK66F18", "STM32F042", "STM32F072", "STM32F103", "STM32F303", "STM32F401", "STM32F407", "STM32F411", "STM32F446", "STM32G431", "STM32G474", "STM32L433", "STM32L443", "atmega16u2", "atmega32u2", "atmega16u4", "atmega32u4", "at90usb162", "at90usb646", "at90usb647", "at90usb1286", "at90usb1287", "atmega32a", "atmega328p", "atmega328", "attiny85", "unknown"]
```

File: lib/python/qmk/constants.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 MAX_KEYBOARD_SUBFOLDERS = 5
 
 # Supported processor types
-CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'MK66F18', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F411', 'STM32F446', 'STM32G431', 'STM32G474', 'STM32L433', 'STM32L443'
+CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'MK66F18', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F407XE', 'STM32F407XG', 'STM32F411', 'STM32F446', 'STM32G431', 'STM32G474', 'STM32L433', 'STM32L443'
```
Comment (tzarc) [inline lib/python/qmk/constants.py:13]:
```suggestion
CHIBIOS_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'MK66F18', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303', 'STM32F401', 'STM32F407', 'STM32F411', 'STM32F446', 'STM32G431', 'STM32G474', 'STM32L433', 'STM32L443'
```

File: platforms/chibios/GENERIC_STM32_F407VG/configs/config.h [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Address for jumping to bootloader on STM32 chips. */
+/* It is chip dependent, the correct number can be looked up by checking against ST's application note AN2606.
+ */
+#define STM32_BOOTLOADER_ADDRESS 0x1FFF77DE
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_F407VG/configs/config.h:20]:
```suggestion
#define STM32_BOOTLOADER_ADDRESS 0x1FFF0000
```

File: platforms/chibios/GENERIC_STM32_F407VG/configs/mcuconf.h [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+#define STM32_SDC_READ_TIMEOUT_MS           1000
+#define STM32_SDC_CLOCK_ACTIVATION_DELAY    10
+#define STM32_SDC_SDIO_UNALIGNED_SUPPORT    TRUE
+#define STM32_SDC_SDIO_DMA_STREAM           STM32_DMA_STREAM_ID(2, 3)
+
+/*
+ * SERIAL driver system settings.
+ */
+#define STM32_SERIAL_USE_USART1             FALSE
+#define STM32_SERIAL_USE_USART2             TRUE
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_F407VG/configs/mcuconf.h:263]:
```suggestion
#define STM32_SERIAL_USE_USART2             FALSE
```

File: quantum/mcu_selection.mk [hunk 7]
```diff
@@ -288,6 +288,78 @@ ifneq ($(findstring STM32F401, $(MCU)),)
...
+
+  # UF2 settings
+  UF2_FAMILY ?= STM32F4
+endif
+
```
Comment (tzarc) [inline quantum/mcu_selection.mk:326]:
```suggestion
```

File: quantum/mcu_selection.mk [hunk 8]
```diff
@@ -288,6 +288,78 @@ ifneq ($(findstring STM32F401, $(MCU)),)
...
+ifneq ($(findstring STM32F407XE, $(MCU)),)
```
Comment (tzarc) [inline quantum/mcu_selection.mk:327]:
```suggestion
ifneq ($(findstring STM32F407, $(MCU)),)
```

File: quantum/mcu_selection.mk [hunk 9]
```diff
@@ -288,6 +288,78 @@ ifneq ($(findstring STM32F401, $(MCU)),)
...
+  #   or <keyboard_dir>/ld/
+  MCU_LDSCRIPT ?= STM32F407xE
+
+  # Startup code to use
+  #  - it should exist in <chibios>/os/common/startup/ARMCMx/compilers/GCC/mk/
+  MCU_STARTUP ?= stm32f4xx
+
+  # Board: it should exist either in <chibios>/os/hal/boards/,
+  # <keyboard_dir>/boards/, or drivers/boards/
+  BOARD ?= GENERIC_STM32_F407VG
```
Comment (tzarc) [inline quantum/mcu_selection.mk:351]:
```suggestion
  BOARD ?= GENERIC_STM32_F407XE
```

**Review by tzarc — CHANGES_REQUESTED** [review]
Also, please move `platforms/chibios/GENERIC_STM32_F407VG` to ` platforms/chibios/GENERIC_STM32_F407XE` -- in general the preference is to provide the "base" version and allow people to override the size through an ldscript override in their rules.mk -- otherwise the core side of things ends up blowing out considerably.

I'll retarget to the `develop` branch because it's core-related, and not a bugfix. Looks like you'll need to rebase -- apologies for that.

File: platforms/chibios/GENERIC_STM32_F407VG/configs/mcuconf.h [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+#define STM32_UART_USART3_DMA_PRIORITY      0
+#define STM32_UART_UART4_DMA_PRIORITY       0
+#define STM32_UART_UART5_DMA_PRIORITY       0
+#define STM32_UART_USART6_DMA_PRIORITY      0
+#define STM32_UART_DMA_ERROR_HOOK(uartp)    osalSysHalt("DMA failure")
+
+/*
+ * USB driver system settings.
+ */
+#define STM32_USB_USE_OTG1                  FALSE
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_F407VG/configs/mcuconf.h:339]: We'll need this, too.
```suggestion
#define STM32_USB_USE_OTG1                  TRUE
```

Comment (tzarc) [reply platforms/chibios/GENERIC_STM32_F407VG/configs/mcuconf.h:339]: Might need to add:
```c
#define STM32_USB_OTG_THREAD_PRIO           NORMALPRIO+1
#define STM32_USB_OTG_THREAD_STACK_SIZE     128
```
...to match the blackpill definitions, too.

File: data/schemas/keyboard.jsonschema (hunk 3 above)
Comment (thpoll83) [reply data/schemas/keyboard.jsonschema:28]: How can you distinguish the 512KB program mem version from the 1MB version? In mcu_selction.mk you have to decide which linker script to use, either STM32F407xG or STM32F407xE

Comment (tzarc) [reply data/schemas/keyboard.jsonschema:28]: By overriding the linker script in the keyboard-specific rules.mk.

Comment (thpoll83) [reply data/schemas/keyboard.jsonschema:28]: Ok, I wasn't aware of that...

**Comment by thpoll83** [comment]
Alright, I rebased to develop and applied all suggestions.

**Comment by tzarc** [comment]
Looks like a couple of cannonkeys files were unintentionally modified -- if we can get that rectified this looks good to go.

File: platforms/chibios/GENERIC_STM32_F407XE/configs/mcuconf.h [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+#define STM32_SDC_READ_TIMEOUT_MS           1000
+#define STM32_SDC_CLOCK_ACTIVATION_DELAY    10
+#define STM32_SDC_SDIO_UNALIGNED_SUPPORT    TRUE
+#define STM32_SDC_SDIO_DMA_STREAM           STM32_DMA_STREAM_ID(2, 3)
+
+/*
+ * SERIAL driver system settings.
+ */
+#define STM32_SERIAL_USE_USART1             FALSE
+#define STM32_SERIAL_USE_USART2             TRUE
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_F407XE/configs/mcuconf.h:263]: Still needs to be set to `FALSE`.

Comment (thpoll83) [reply platforms/chibios/GENERIC_STM32_F407XE/configs/mcuconf.h:263]: sorry, I thought I fixed that...

Comment (drashna) [reply platforms/chibios/GENERIC_STM32_F407XE/configs/mcuconf.h:263]:
```suggestion
#define STM32_SERIAL_USE_USART2             FALSE
```

**Comment by tzarc** [comment]
> Can you please check the PR again?

USART2 change still hasn't been applied, I'm afraid.

File: platforms/chibios/GENERIC_STM32_F407XE/configs/mcuconf.h [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+#define STM32_UART_USART3_DMA_PRIORITY      0
+#define STM32_UART_UART4_DMA_PRIORITY       0
+#define STM32_UART_UART5_DMA_PRIORITY       0
+#define STM32_UART_USART6_DMA_PRIORITY      0
+#define STM32_UART_DMA_ERROR_HOOK(uartp)    osalSysHalt("DMA failure")
+
+/*
+ * USB driver system settings.
+ */
+#define STM32_USB_USE_OTG1                  FALSE
```
Comment (tzarc) [inline platforms/chibios/GENERIC_STM32_F407XE/configs/mcuconf.h:339]:
```suggestion
#define STM32_USB_USE_OTG1                  TRUE
```

**Review by tzarc — APPROVED** [review]

**Review by zvecr — APPROVED** [review]

**Pull request merged**